[SourceKit] Use a single PluginRegistry in multiple ASTContexts

rintaro opened the pull request:
Make a single `PluginRegistry` and share it between `SwiftASTManager`, `IDEInspectionInstance`, and `CompileInstance`. And inject the plugin registry to `ASTContext` right after `CompilerInstance.setup()`

That way, all sema-capable `ASTContext` in SourceKit share a single `PluginRegistry`.

rintaro committed:
[SourceKit] Use a single PluginRegistry in multiple ASTContexts

Make a single 'PluginRegistry' and share it between SwiftASTManager,
IDEInspectionInstance, and CompileInstance. And inject the plugin
registry to ASTContext right after 'CompilerInstance.setup()'

That way, all sema-capable ASTContext in SourceKit share a single
PluginRegistry.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test macOS

DougGregor commented on include/swift/AST/PluginRegistry.h:
```diff
@@ -139,6 +139,8 @@ class PluginRegistry {
   /// Flag to dump plugin messagings.
   bool dumpMessaging = false;
 
+  std::mutex mtx;
```
`getLoadedLibraryPlugins()` needs to lock this, too.

rintaro replied:
I added a commit in this PR removing `PluginRegistry::getLoadedLibraryPlugins()`.

rintaro committed:
[Macros] Track loaded plugin paths in each ASTContext

PluginRegistry is now shared between multiple ASTContext. ASTContext
should track its loaded plugin library paths separately.

rintaro commented:
@swift-ci Please smoke test

rintaro commented on lib/AST/ASTContext.cpp:
```diff
@@ -6395,6 +6397,9 @@ LoadedExecutablePlugin *ASTContext::loadExecutablePlugin(StringRef path) {
 }
 
 void *ASTContext::loadLibraryPlugin(StringRef path) {
+  // Remember the path (even if it fails to load.)
+  getImpl().LoadedPluginLibraryPaths.insert(path);
```
Do we want to track _failed_ libraries? If not, I can just move this after `getPluginRegistry()->loadLibraryPlugin(resolvedPath)` .

Also, do we want the VFS path or the "real" path here? I guess it doesn't matter because we only use this to get the module name from the filename. cc: @rxwei
